designs/virtualization/4.x/create-vm-dropdowns/create-vm-dropdowns.md

```diff
@@ -0,0 +1,112 @@
+---
+parent: Virtualization
+version: 4.x
+---
+
+# create VM dropdowns
+
+This item will cover various updates to the VM creation wizard
+- Adding descriptions to the dropdown fields (using this pattern https://github.com/patternfly/patternfly-design-kit/issues/218)
+- Renaming fields
+- Reordering fields
+
+
+![Create VM - General-default](img/1-0.jpg)
+Landing on the General step, Default mode.
+Section titles added, and fields reorgenized and renamed.
+'General' section:
+- 'Name'
+- 'Description'
+'Resources & configuration' section'
+- 'Operating system'
+- 'Configuration' (previously Workload profile)
+- 'Flavor (CPU and Memory)' (previously only 'flavor')
+'Operating system image source' section
+- 'Boot operating system via' (previously 'source')
+
+## General step
+
+![Create VM - General](img/1-1.jpg)
+Upon selecting an operating system, the configuration and Flavor (CPU and Memory) fields will be enabled and their defaults will be selected.
+
+![Create VM - General - VM configuration](img/1-1-1.jpg)
+VM Configuration options
+Server is the default
+
+![Create VM - General - flavor](img/1-1-2.jpg)
+Flavor options
+Medium is the default
+
+![Create VM - General](img/1-2-1.jpg)
+'Boot operating system via' field options
+
+![Create VM - General](img/1-2.jpg)
+Selecting an option from the 'Boot operating system via' field will expose below all the fields required to connect the VM to an OS image source.
```

Comment:
@suomiy 
The point of this effort is to avoid making the user complete the same task in different steps.
The user is facing a difficult task here, so by splitting it to two different screens will not remove complexity, only add to it.
This task must be done anyway, so it is better experienced to complete it at one go, while the user is in focus and flow.
The Review step should not change, all the NIC/Disk data will be displayed on the correlating Network/Storage steps, same as it is currently implemented.

Reply:
@suomiy 
Sorry, I don't see any workaround.

I tried to use the same Disk/Nic modal only from the general step, the context was lost again since there is no CTA to actually "Select OS image" within the modal. So to make it work, we do have to make some customization like the name of the modal and blocking the "Blank" option form the Disk content field. Take a look (@matthewcarleton let me know what you think too): 
![1-1 copy A](https://user-images.githubusercontent.com/48942365/84006183-f92e0800-a976-11ea-9da0-06e9ab1d3f8a.jpg)
![1-1 copy B](https://user-images.githubusercontent.com/48942365/84006187-fc28f880-a976-11ea-8400-6ffecbe2cd65.jpg)
![1-1 copy C](https://user-images.githubusercontent.com/48942365/84006189-fd5a2580-a976-11ea-9879-b491ab438762.jpg)

I also compared this design to the "Complete this step on the Network/Storage step". Pushing this entire process to the General step works better for the same reasons.

Reply:
it feels a bit strange to select New PVC here and then show a table where they need to click an action to get to a modal to interact with that modal. It would be ideal (if we are saying we need the modal to pop up) for the user to have that option from the dropdown. It feels a bit strange though I think to open a modal from a dropdown. 
One thing we are missing here too is the upload right? We should at the very least describe it as an available action otherwise we must assume the user doesn't know it is available to them.

Reply:
> The point of this effort is to avoid making the user complete the same task in different steps.
The user is facing a difficult task here, so by splitting it to two different screens will not remove complexity, only add to it.
This task must be done anyway, so it is better experienced to complete it at one go, while the user is in focus and flow.

IMO it seemed to help the VM Import flow quite a bit - moving the provider stuff into its own tab.
There might be too much happening on one page and maybe it would make more sense to focus on one task instead (picking OS image).

I guess we will never know, until we get some feedback on this from users.

> I tried to use the same Disk/Nic modal only from the general step, the context was lost again since there is no CTA to actually "Select OS image" within the modal. So to make it work, we do have to make some customization like the name of the modal and blocking the "Blank" option form the Disk content field. 

It should be quite feasible to customize the modal for the OS source selection.

> it feels a bit strange to select New PVC here and then show a table where they need to click an action to get to a modal to interact with that modal. It would be ideal (if we are saying we need the modal to pop up) for the user to have that option from the dropdown. It feels a bit strange though I think to open a modal from a dropdown.

I agree, it is probably unnecessary to show the whole table for just one entity.

>One thing we are missing here too is the upload right? We should at the very least describe it as an available action otherwise we must assume the user doesn't know it is available to them.

It should be possible to change the source in the modal to upload as well. Although not sure on how will @glekner end up implementing it and how it will impact the whole wizard flow

Reply:
> it feels a bit strange to select New PVC here and then show a table where they need to click an action to get to a modal to interact with that modal.

Agree

> It would be ideal (if we are saying we need the modal to pop up) for the user to have that option from the dropdown. It feels a bit strange though I think to open a modal from a dropdown.
One thing we are missing here too is the upload, right? We should at the very least describe it as an available action otherwise we must assume the user doesn't know it is available to them.
> It should be quite feasible to customize the modal for the OS source selection.
This supports the original design suggestion. Anyway, the design I added in this comment is not working (showing list within the general step and all). The CTA to create a new disk with a button fees funny to me, it is out of sync with the page flow. Here's another implementation, but I still think in-page is better:
![1-1 copy Upload local file Copy](https://user-images.githubusercontent.com/48942365/84130476-01a44280-aa4c-11ea-891a-13031d164919.jpg)


I think that all of this supports that it would be easier to surface all the required fields in-page/line without the need to customize any modal.

> IMO it seemed to help the VM Import flow quite a bit - moving the provider stuff into its own tab.

I think it makes sense to add the OS image as a new step.

@suomiy 
I admit that I don't see the issue with the in-page/line solution. It is already implemented today.
Today, when the user selects PXE, URL, and Container options, we already follow-up with the required fields in-page/line below the OS image field. So why not align this existing behavior to include other options available from the OS image as well? Why break "Existing PVC/ Clone an existing PVC/New PVC (upload) from this flow?

Reply:
> I think that all of this supports that it would be easier to surface all the required fields in-page/line without the need to customize any modal.

It is quite easy to customize the modal. But adding a subset of inline fields to the page will make it quite a pain. Disparity of functionality/validations  + two places for maintanance/bugs

> I think it makes sense to add the OS image as a new step.

+1, this sounds good as it doesn't have the problems mentioned above (the inlining part). We could reuse full feature disk modal with few changes in the step variant.

> Today, when the user selects PXE, URL, and Container options, we already follow-up with the required fields in-page/line below the OS image field. So why not align this existing behavior to include other options available from the OS image as well?

As I mentioned - the features/validations. We are not even enforcing the required part today.

> Why break "Existing PVC/ Clone an existing PVC/New PVC (upload) from this flow?

Because, there are many fields/validations which are relevant for the user

Anyway we should take boot order and disk removal into considerations as well. These will change the source outside of this flow.

Reply:
Also the current iteration with Add disk button which will opens the disk dialog looks quite fine.

Reply:
I've updated the PR to include the design we all found favorite (creating a new "OS source" step)